circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/JceksAmazonS3ClientFactory.java

```diff
@@ -38,20 +41,43 @@
   private final static Logger LOG = LoggerFactory.getLogger(JceksAmazonS3ClientFactory.class);
 
   private final Security security;
+  private final HiveConf conf;
 
   @Autowired
-  public JceksAmazonS3ClientFactory(Security security) {
+  public JceksAmazonS3ClientFactory(Security security, HiveConf replicaHiveConf) {
     this.security = security;
+    conf = replicaHiveConf;
+  }
+
+  public JceksAmazonS3ClientFactory(Security security) {
+    this(security, null);
   }
 
   @Override
   public AmazonS3 newInstance(AmazonS3URI uri, S3S3CopierOptions s3s3CopierOptions) {
+    HadoopAWSCredentialProviderChain credentialProviderChain = getCredentialsProviderChain(
+        s3s3CopierOptions.getAssumedRole());
+    return newS3Client(uri, s3s3CopierOptions, credentialProviderChain);
+  }
+
+  private AmazonS3 newS3Client(
+      AmazonS3URI uri,
+      S3S3CopierOptions s3s3CopierOptions,
+      HadoopAWSCredentialProviderChain credentialProviderChain) {
     LOG.debug("trying to get a client for uri '{}'", uri);
-    AmazonS3 globalClient = newGlobalInstance(s3s3CopierOptions);
+    AmazonS3 globalClient = newGlobalInstance(s3s3CopierOptions, credentialProviderChain);
     try {
+
+      /*
+       * When using roles it can take a while for the credentials to be retrieved from the
+       * AssumeRoleCredentialsProvider. This can mean that the rest of the code completes before the credentials are
+       * retrieved, resulting in errors. A temporary fix for this situation is to put the thread to sleep for 10s to
```

Comment:
Where is this sleep happening? Is it not part of this PR, just a recommendation?

Reply:
It's not currently happening - we added it in to fix the issue mentioned but then changed the code and found it was no longer needed. It was suggested that we keep it in a comment just in case any one comes across the issue again. It's a bit of a weird one though, and the comment seems to be confusing a few people so unsure if its all that helpful.